File: 1billionRowsProblem/read1billionRows.py
import multiprocessing
import time
import sys
import math

def process_chunk(chunk):
    """
    Process a chunk of data (a list of dict items) and return aggregated results.
    """
    local_map = {}
    for data in chunk:
        for city, tmp in data.items():
            if city in local_map:
                city_data = local_map[city]
                count = city_data['count']
                city_data['count'] += 1
                city_data['avg'] = (city_data['avg'] * count + tmp) / city_data['count']
                city_data['min'] = min(city_data['min'], tmp)
                city_data['max'] = max(city_data['max'], tmp)
            else:
                local_map[city] = {'min': tmp, 'max': tmp, 'avg': tmp, 'count': 1}
    return local_map

def merge_results(map1, map2):
    """
    Merge two aggregated result dictionaries.
    """
    for city, data in map2.items():
        if city in map1:
            data1 = map1[city]
            total_count = data1['count'] + data['count']
            new_avg = ((data1['avg'] * data1['count']) + (data['avg'] * data['count'])) / total_count
            map1[city] = {
                'min': min(data1['min'], data['min']),
                'max': max(data1['max'], data['max']),
                'avg': new_avg,
                'count': total_count
            }
        else:
            map1[city] = data
    return map1

def read_input_file_in_chunks(filename, chunk_size=100000):
    """
    Generator that yields chunks (lists) of dictionaries from the file.
    Each line in the file should be in the format: City:Value
    """
    with open(filename, 'r') as f:
        chunk = []
        for line in f:
            line = line.strip()
            if not line:
                continue
            parts = line.split(':')
            if len(parts) != 2:
                continue
            city, tmp = parts
            try:
                data = {city.strip(): float(tmp.strip())}
            except ValueError:
                continue
            chunk.append(data)
            if len(chunk) >= chunk_size:
                yield chunk
                chunk = []
        if chunk:
            yield chunk

def count_lines(filename):
    """
    Counts the total number of lines in the input file.
    """
    with open(filename, 'r') as f:
        return sum(1 for _ in f)

def get_final_data(updated_data_map):
    """
    Rounds the aggregated values to two decimal places (excluding 'count').
    """
    temperature_data = {}
    for city, tmp_data in updated_data_map.items():
        temperature_data[city] = {key: round(value, 2) for key, value in tmp_data.items() if key != 'count'}
    return temperature_data

if __name__ == '__main__':
    filename = '1billionRowInput.txt'
    chunk_size = 100000
    start_time = time.time()

    # Count total lines and calculate the number of chunks for progress reporting.
    total_lines = count_lines(filename)
    total_chunks = math.ceil(total_lines / chunk_size)
    chunks_processed = 0

    final_map = {}
    num_processes = multiprocessing.cpu_count()
    print(f'{num_processes=}')
    pool = multiprocessing.Pool(processes=num_processes)

    # Process each chunk in parallel using imap
    for partial_map in pool.imap(process_chunk, read_input_file_in_chunks(filename, chunk_size)):
        final_map = merge_results(final_map, partial_map)
        chunks_processed += 1
        percentage = (chunks_processed / total_chunks) * 100
        # Update progress on the same line
        sys.stdout.write("\rProgress: {:.2f}%".format(percentage))
        sys.stdout.flush()

    pool.close()
    pool.join()

    # Final rounding for the aggregated results
    final_data = get_final_data(final_map)
    print("\nFinal aggregated data:")
    print(final_data)
    print("Processed in: {:.2f} seconds".format(time.time() - start_time))

# Processed in: 372.03 seconds

File: 1billionRowsProblem/read1billionRows.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

// CityData stores the aggregated values for each city.
type CityData struct {
	Min, Max, Avg float64
	Count         int
}

// processChunk processes a slice of lines and computes the min, max, average, and count per city.
func processChunk(chunk []string) map[string]CityData {
	localMap := make(map[string]CityData)
	for _, line := range chunk {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		parts := strings.Split(line, ":")
		if len(parts) != 2 {
			continue
		}
		city := strings.TrimSpace(parts[0])
		tmpStr := strings.TrimSpace(parts[1])
		tmp, err := strconv.ParseFloat(tmpStr, 64)
		if err != nil {
			continue
		}
		if data, ok := localMap[city]; ok {
			newCount := data.Count + 1
			// Update average using a weighted average
			data.Avg = (data.Avg*float64(data.Count) + tmp) / float64(newCount)
			if tmp < data.Min {
				data.Min = tmp
			}
			if tmp > data.Max {
				data.Max = tmp
			}
			data.Count = newCount
			localMap[city] = data
		} else {
			localMap[city] = CityData{Min: tmp, Max: tmp, Avg: tmp, Count: 1}
		}
	}
	return localMap
}

// mergeResults merges two maps of aggregated city data.
func mergeResults(map1, map2 map[string]CityData) map[string]CityData {
	for city, data2 := range map2 {
		if data1, exists := map1[city]; exists {
			totalCount := data1.Count + data2.Count
			newAvg := ((data1.Avg * float64(data1.Count)) + (data2.Avg * float64(data2.Count))) / float64(totalCount)
			mergedData := CityData{
				Min:   math.Min(data1.Min, data2.Min),
				Max:   math.Max(data1.Max, data2.Max),
				Avg:   newAvg,
				Count: totalCount,
			}
			map1[city] = mergedData
		} else {
			map1[city] = data2
		}
	}
	return map1
}

// countLines counts the total number of lines in the given file.
func countLines(filename string) (int, error) {
	file, err := os.Open(filename)
	if err != nil {
		return 0, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	count := 0
	for scanner.Scan() {
		count++
	}
	return count, scanner.Err()
}

func main() {
	filename := "1billionRowInput.txt"
	chunkSize := 100000
	startTime := time.Now()

	// Count total lines to determine total chunks (for progress reporting)
	totalLines, err := countLines(filename)
	if err != nil {
		fmt.Println("Error counting lines:", err)
		return
	}
	totalChunks := int(math.Ceil(float64(totalLines) / float64(chunkSize)))

	// Open the file for processing.
	file, err := os.Open(filename)
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var (
		wg              sync.WaitGroup
		resultsChan     = make(chan map[string]CityData)
		mu              sync.Mutex
		finalMap        = make(map[string]CityData)
		chunksProcessed = 0
	)

	// Launch a goroutine to update progress on the same console line.
	go func() {
		for {
			mu.Lock()
			cp := chunksProcessed
			mu.Unlock()
			percentage := (float64(cp) / float64(totalChunks)) * 100
			fmt.Printf("\rProgress: %.2f%%", percentage)
			if cp >= totalChunks {
				break
			}
			time.Sleep(100 * time.Millisecond)
		}
	}()

	// Read the file line by line and group them into chunks.
	var lines []string
	for scanner.Scan() {
		line := scanner.Text()
		lines = append(lines, line)
		if len(lines) >= chunkSize {
			// Copy the current chunk to avoid data race.
			chunk := make([]string, len(lines))
			copy(chunk, lines)
			wg.Add(1)
			go func(ch []string) {
				defer wg.Done()
				res := processChunk(ch)
				resultsChan <- res
			}(chunk)
			lines = nil // reset for the next chunk
		}
	}
	// Process any remaining lines as the final chunk.
	if len(lines) > 0 {
		wg.Add(1)
		chunk := make([]string, len(lines))
		copy(chunk, lines)
		go func(ch []string) {
			defer wg.Done()
			res := processChunk(ch)
			resultsChan <- res
		}(chunk)
	}

	// Close the results channel when all chunks have been processed.
	go func() {
		wg.Wait()
		close(resultsChan)
	}()

	// Merge the results from each processed chunk.
	for partial := range resultsChan {
		finalMap = mergeResults(finalMap, partial)
		mu.Lock()
		chunksProcessed++
		mu.Unlock()
	}

	// Ensure progress shows 100%
	fmt.Printf("\rProgress: 100.00%%\n")

	// Round the final aggregated values to 2 decimals.
	finalData := make(map[string]map[string]float64)
	for city, data := range finalMap {
		finalData[city] = map[string]float64{
			"min": math.Round(data.Min*100) / 100,
			"max": math.Round(data.Max*100) / 100,
			"avg": math.Round(data.Avg*100) / 100,
		}
	}

	fmt.Println("Final aggregated data:")
	fmt.Println(finalData)
	fmt.Printf("Processed in: %.2f seconds\n", time.Since(startTime).Seconds())
}

// Processed in: 31.46 seconds
